operator/pkg/util/apiclient/idempotency.go

```diff
@@ -0,0 +1,222 @@
+		if err != nil {
+			return err
+		}
+
+		mwc.ResourceVersion = older.ResourceVersion
+		_, err = client.AdmissionregistrationV1().MutatingWebhookConfigurations().Update(context.TODO(), mwc, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	klog.V(5).InfoS("Successfully created or updated mutatingWebhookConfiguration", "mutatingWebhookConfiguration", mwc.GetName())
+	return nil
+}
+
+// CreateOrUpdateValidatingWebhookConfiguration creates a ValidatingWebhookConfiguration if the target resource doesn't exist. If the resource exists already, this function will update the resource instead.
+func CreateOrUpdateValidatingWebhookConfiguration(client clientset.Interface, vwc *admissionregistrationv1.ValidatingWebhookConfiguration) error {
+	_, err := client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Create(context.TODO(), vwc, metav1.CreateOptions{})
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+		older, err := client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Get(context.TODO(), vwc.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		vwc.ResourceVersion = older.ResourceVersion
+		_, err = client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Update(context.TODO(), vwc, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	klog.V(5).InfoS("Successfully created or updated validatingWebhookConfiguration", "validatingWebhookConfiguration", vwc.GetName())
+	return nil
+}
+
+// CreateOrUpdateAPIService creates a APIService if the target resource doesn't exist. If the resource exists already, this function will update the resource instead.
+func CreateOrUpdateAPIService(apiRegistrationClient *aggregator.Clientset, apiservice *apiregistrationv1.APIService) error {
+	_, err := apiRegistrationClient.ApiregistrationV1().APIServices().Create(context.TODO(), apiservice, metav1.CreateOptions{})
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+		older, err := apiRegistrationClient.ApiregistrationV1().APIServices().Get(context.TODO(), apiservice.GetName(), metav1.GetOptions{})
+		if err != nil {
+			return err
+		}
+
+		apiservice.ResourceVersion = older.ResourceVersion
+		_, err = apiRegistrationClient.ApiregistrationV1().APIServices().Update(context.TODO(), apiservice, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	klog.V(5).Infof("Successfully created or updated APIService", "APIService", apiservice.Name)
+	return nil
+}
+
+// CreateOrUpdateCustomResourceDefinition creates a CustomResourceDefinition if the target resource doesn't exist. If the resource exists already, this function will update the resource instead.
+func CreateOrUpdateCustomResourceDefinition(client *crdsclient.Clientset, obj *apiextensionsv1.CustomResourceDefinition) error {
```

Comment:
Here, I think that if there already have karmada crd in karmada controlplane, we will skip create crd, because the crd content is invariable.
 Can we change the func name `CreateOrUpdateCustomResourceDefinition` to `CreateCustomResourceDefinitionIfNeed` ?

Reply:
Do we need to consider that the crds are deleted by mistake?